File: Java/491. Non-decreasing Subsequences.java
class Solution {
    Set<List<Integer>> ans;

    public List<List<Integer>> findSubsequences(int[] nums) {
        ans = new HashSet();
        backtrack(nums, 0, new ArrayList());
        return new ArrayList(ans);
    }

    private void backtrack(int[] nums, int i, List<Integer> path) {
        if (i == nums.length) {
            if (path.size() >= 2) {
                ans.add(new ArrayList(path));
            }
            return;
        }

        if (path.isEmpty() || path.get(path.size() - 1) <= nums[i]) {
            path.add(nums[i]);
            backtrack(nums, i + 1, path);
            path.remove(path.size() - 1);
        }

        backtrack(nums, i + 1, path);
    }
}

File: Python3/491. Non-decreasing Subsequences.py
class Solution:
    def findSubsequences(self, nums: List[int]) -> List[List[int]]:
        ans = set()
        path = []

        def backtrack(i = 0):

            if i == len(nums):
                if len(path) > 1:
                    ans.add(tuple(path))
                return

            if not path or path[-1] <= nums[i]:
                path.append(nums[i])
                backtrack(i + 1)
                path.pop()
            
            backtrack(i + 1)

        backtrack()
        return ans